### PR TITLE
Fixes sending of rst_stream

### PR DESCRIPTION
**Pull request by tbetbetbe**
- rst_stream was not being sent unless there was a protocol error
- tests were broken by this change:
  * so fixed the broken tests by renaming StreamError -> InternalError
  * This was necessary because
    + stream#stream_error defaulted to calling close with :stream_error
    + :stream_error is not a DEFINED_ERROR, and could not be converted
      into a valid exception class

### CODE REVIEW

File: lib/http/2/connection.rb [hunk 1]
```diff
@@ -342,7 +342,12 @@ def send(frame)
         # An endpoint can end a connection at any time. In particular, an
         # endpoint MAY choose to treat a stream error as a connection error.
         if frame[:type] == :rst_stream
-          goaway(frame[:error]) if frame[:error] == :protocol_error
```
Comment (igrigorik) [inline lib/http/2/connection.rb]: This duplicates same code as else clause. We can make this cleaner.

File: lib/http/2/error.rb [hunk 2]
```diff
@@ -25,7 +25,7 @@ class FlowControlError < ProtocolError; end
 
     # Raised on invalid stream processing: invalid frame type received or
     # sent, or invalid command issued.
-    class StreamError < ProtocolError; end
+    class InternalError < ProtocolError; end
```
Comment (igrigorik) [inline lib/http/2/error.rb]: Why is this necessary? StreamError is a much better name indicating the actual type of error.

Comment (tbetbetbe) [inline lib/http/2/error.rb]: From Stream#stream_error
If stream_error is supposed to be a defined error, then we can add it an remove this change, I simply assumed it was not.

```ruby

    def stream_error(error = :stream_error, msg: nil)
      @error = error
      close(error) if @state != :closed   # Here, IUC, error needs to be a valid defined error

      klass = error.to_s.split('_').map(&:capitalize).join
      fail Error.const_get(klass), msg  # Here, the Exception class name needs to match the error name
    end
```

Comment (igrigorik) [inline lib/http/2/error.rb]: Ah, ok, I'm with you now -- lgtm.

**Comment by igrigorik** [comment]
@tbetbetbe can we eliminate the duplication in https://github.com/igrigorik/http-2/pull/38#discussion_r36664031?

@mad-p any thoughts on this one?

File: lib/http/2/connection.rb (hunk 1 above)
Comment (tbetbetbe) [inline lib/http/2/connection.rb]: Fixed

**Comment by igrigorik** [comment]
:+1: thanks Tim!

File: spec/stream_spec.rb [hunk 3]
```diff
@@ -56,13 +56,15 @@
...
+          expect { @stream.dup.send type }.to raise_error InternalError
         end
       end
 
       it 'should raise error on receipt of invalid frames' do
-        (FRAME_TYPES - [PRIORITY, RST_STREAM, WINDOW_UPDATE]).each do |type|
-          expect { @stream.dup.receive type }.to raise_error StreamError
+        what_types = (FRAME_TYPES - [PRIORITY, RST_STREAM, WINDOW_UPDATE])
+        what_types.each do |type|
+          p "test type #{type}"
```
Comment (mad-p) [inline spec/stream_spec.rb]: Debug print remained

**Comment by mad-p** [comment]
Thanks, Tim! lgtm
